jgrapht-core/src/main/java/org/jgrapht/alg/isomorphism/AHUTreeIsomorphismInspector.java

```diff
@@ -0,0 +1,349 @@
+
+        this.root1 = Objects.requireNonNull(root1, "root1 cannot be null");
+        this.root2 = Objects.requireNonNull(root2, "root2 cannot be null");
+
+        if (!tree1.vertexSet().contains(root1)){
+            throw new IllegalArgumentException("root not contained in forest");
+        }
+
+        if (!tree2.vertexSet().contains(root2)){
+            throw new IllegalArgumentException("root not contained in forest");
+        }
+    }
+
+    private void bfs(Graph<V, E> graph, V root, List<List<V>> levels){
+        BreadthFirstIterator<V, E> bfs = new BreadthFirstIterator<>(graph, root);
+
+        while (bfs.hasNext()){
+            V u = bfs.next();
+
+            if (levels.size() < bfs.getDepth(u) + 1){
+                levels.add(new ArrayList<>());
+            }
+
+            levels.get(bfs.getDepth(u)).add(u);
+        }
+    }
+
+    private List<List<V>> computeLevels(Graph<V, E> graph, V root){
+        List<List<V>> levels = new ArrayList<>();
+        bfs(graph, root, levels);
+        return levels;
+    }
+
+    private void matchVerticesWithSameLabel(V root1, V root2,
+                                            Map<V, Integer>[] canonicalName){
+        Queue<Pair<V, V>> queue = new ArrayDeque<>();
+        queue.add(Pair.of(root1, root2));
+
+        while (!queue.isEmpty()){
+            Pair<V, V> pair = queue.poll();
+            V u = pair.getFirst();
+            V v = pair.getSecond();
+
+            forwardMapping.put(u, v);
+            backwardMapping.put(v, u);
+
+            Map<Integer, List<V>> labelList = new HashMap<>(tree1.degreeOf(u));
+
+            for (E edge: tree1.edgesOf(u)){
+                V next = Graphs.getOppositeVertex(tree1, edge, u);
+
+                if (!forwardMapping.containsKey(next)){
+                    labelList.computeIfAbsent(canonicalName[0].get(next), x -> new ArrayList<>()).add(next);
+                }
+            }
+
+            for (E edge: tree2.edgesOf(v)){
+                V next = Graphs.getOppositeVertex(tree2, edge, v);
+
+                if (!backwardMapping.containsKey(next)){
+                    List<V> list = labelList.get(canonicalName[1].get(next));
+
+                    if (list == null || list.isEmpty()){
+                        forwardMapping.clear();
+                        backwardMapping.clear();
+                        return;
+                    }
+
+                    V pairedNext = list.remove(list.size() - 1);
+                    queue.add(Pair.of(pairedNext, next));
+                }
+            }
+        }
+    }
+
+    private boolean isomorphismExists(V root1, V root2){
+        this.forwardMapping = new HashMap<>();
+        this.backwardMapping = new HashMap<>();
+
+        // Are both graphs empty?
+        if (Objects.isNull(root1) && Objects.isNull(root2))
+            return true;
+
+        @SuppressWarnings("unchecked")
+        Map<V, Integer>[] canonicalName = new Map[2];
```

Comment:
When you know the class, you use the `Array.newInstance(Map.class, 2)` to create generic arrays.